stl/inc/flat_map

```diff
@@ -0,0 +1,885 @@
+        _Data.keys        = _STD move(_Key_cont);
+        _Data.values      = _STD move(_Mapped_cont);
+        _Guard._Clearable = nullptr;
+    }
+
+    // observers
+    key_compare key_comp() const {
+        return _Key_compare;
+    }
+
+    value_compare value_comp() const {
+        return value_compare(_Key_compare);
+    }
+
+    const key_container_type& keys() const noexcept {
+        return _Data.keys;
+    }
+
+    const mapped_container_type& values() const noexcept {
+        return _Data.values;
+    }
+
+    // map operations
+    template <class _K>
+    iterator find(const _K& _X)
+        requires same_as<_K, key_type> || _Compare_is_transparent<key_compare>
+    {
+        iterator _Position = lower_bound(_X);
+        if (_Position != end() && _Key_equal(_Position->first, _X)) {
+            return _Position;
+        } else {
+            return end();
+        }
+    }
+
+    template <class _K>
+    const_iterator find(const _K& _X) const
+        requires same_as<_K, key_type> || _Compare_is_transparent<key_compare>
+    {
+        const_iterator _Position = lower_bound(_X);
+        if (_Position != cend() && _Key_equal(_Position->first, _X)) {
+            return _Position;
+        } else {
+            return cend();
+        }
+    }
+
+    template <class _K>
+    size_type count(const _K& _X) const
+        requires same_as<_K, key_type> || _Compare_is_transparent<key_compare>
+    {
+        return upper_bound(_X) - lower_bound(_X);
+    }
+
+    template <class _K>
+    bool contains(const _K& _X) const
+        requires same_as<_K, key_type> || _Compare_is_transparent<key_compare>
+    {
+        return find(_X) != cend();
+    }
+
+    template <class _K>
+    iterator lower_bound(const _K& _X)
+        requires same_as<_K, key_type> || _Compare_is_transparent<key_compare>
+    {
+        typename KeyContainer::const_iterator _Key_it =
+            _STD lower_bound(_Data.keys.cbegin(), _Data.keys.cend(), _X, _Key_compare);
+        auto _Dist                                 = _STD distance(_Data.keys.cbegin(), _Key_it);
+        typename MappedContainer::iterator _Val_it = _Data.values.begin() + _Dist;
+        return iterator{_STD move(_Key_it), _STD move(_Val_it)};
+    }
+
+    template <class _K>
+    const_iterator lower_bound(const _K& _X) const
+        requires same_as<_K, key_type> || _Compare_is_transparent<key_compare>
+    {
+        typename KeyContainer::const_iterator _Key_it =
+            _STD lower_bound(_Data.keys.cbegin(), _Data.keys.cend(), _X, _Key_compare);
+        auto _Dist                                       = _STD distance(_Data.keys.cbegin(), _Key_it);
+        typename MappedContainer::const_iterator _Val_it = _Data.values.cbegin() + _Dist;
+        return const_iterator{_STD move(_Key_it), _STD move(_Val_it)};
+    }
+
+    template <class _K>
+    iterator upper_bound(const _K& _X)
+        requires same_as<_K, key_type> || _Compare_is_transparent<key_compare>

```

Comment:
Perhaps we need separate `upper_bound(const key_type&)` and `upper_bound(const auto&) requires (_Compare_is_transparent<key_compare>)` functions. For this and other similar methods.
The following example gives some idea what happens with one single definition. 
```cpp
#include<type_traits>

constexpr bool _Allowed = false;

#if 0
template<class T>
    requires std::is_same_v<T, int> || _Allowed
void func(const T&) {
}
// no. `T==double`, but both `is_same_v<double,int>` and `_Allowed` are false.
void test() {
    func(1.0);
}
#else
void func(const int&) {
}
template<class T>
    requires _Allowed
void func(const T&) {
}
// ok. selects func(const int&).
void test() {
    func(1.0);
}
#endif

int main() {

}
```

Reply:
Do you think it can be fixed by changing `same_as` to `convertible_to`?

Reply:
There are cases `convertible_to` doesn't work either. As `upper_bound` takes both `_K:_X` and `_Key_compare`. If `_Key_compare` rejects types other than `key_type` then this also fail to compile.

Reply:
OK. I will fix this later. I'll need to learn more about overload resolution. I want to find a way to avoid duplicating the function twice.

Reply:
It seems to me that we have to make the conversion to the `key_type` at the call site in some cases if such conversion is necessary ([Godbolt link](https://godbolt.org/z/W85KT5Ec4)), and thus separated overloads are needed.

The forwarding doesn't work (in other words, can't be perfect) when
- `key_type` is implicitly convertible from class `S`, and
- the conversion is estabilished by constructor which takes `S` by value, and
- `S` is neither movable nor copyable.

In such a case, the conversion to `key_type` has to be done outside of `upper_bound` if one calls `m.upper_bound(S{/* args* /})`.

Reply:
I think I solved this. Please take a look again.